### PR TITLE
Amazon Publisher Services Pull Request

### PR DESCRIPTION
**Pull request by AmazonPublisherServices**
Submitting initial pull request for amp-ad implementation on behalf of Amazon Publisher Services. 

From intent to implement #12339. 

We're excited to become part of the AMP Project! Looking forward to your feedback!

### CODE REVIEW

**Comment by AmazonPublisherServices** [comment]
Seeing this error from travis, this is not a bug, it is intended to be written this way

Error compiling ./3p/integration.js
ads/aps.js:58: WARNING - Suspicious code. This code lacks side-effects. Is there a bug?
      data.bidTimeout ? delete data.bidTimeout : null;
                                                 ^^^^

**Comment by lannka** [comment]
@AmazonPublisherServices thanks for the work. could you please implement the render-start and no-content APIs? check out [here](https://github.com/ampproject/amphtml/blob/master/ads/README.md#other-tips)

Regarding to he travis failure, can you do the "delete" in the 2nd line to please the linter check?

**Comment by AmazonPublisherServices** [comment]
@lannka Absolutely! Will work on these and submit again soon!

**Comment by AmazonPublisherServices** [comment]
@lannka After looking at the documentation, we would need to integrate the APIs inside of our library, which we can do, but because we render inside of a second cross domain iframe, we would also need to use the useSameDomainRenderingUntilDeprecated feature so we can access the methods provided by the amp runtime. 

Is this required to begin the merge process?

**Comment by lannka** [comment]
is your second iframe trying to make an ad request? or is it to render an ad response?
if it is to render, then you should be able to call renderStart right before creating the iframe.

`useSameDomainRenderingUntilDeprecated` is not relevant here (it's a doubleclick feature).

**Comment by AmazonPublisherServices** [comment]
@lannka the second iframe is rendering the ad response, but we do not have access to the APIs because because our render function is called from the DFP creative, which is inside of the second cross domain iframe.

**Comment by lannka** [comment]
Sorry, a bit confused. Is it your script who got the ad response? Or are you inserting an xorigin iframe that will make an ad request?

**Comment by AmazonPublisherServices** [comment]
So the request goes out in ads/aps.js, our response comes back and we pass some targeting values into doubleclick (all in ads/aps.js). If our line item is selected in DFP, the DFP creative iframe serves our render function, which I believe is why we don't have access to those APIs. Does that make sense? Have you run into this with any other bidders who pass targeting values into doubleclick?

**Comment by lannka** [comment]
@AmazonPublisherServices I see. So your server does not return ad, but some targeting values.

That is fine. I can forget about the API for now.

File: ads/aps.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      const bidObject = bids.filter(bid => bid.slotID === data.slot)[0];
+      if (bidObject) {
+        global.apstag.targetingKeys('ampDisplay')
+            .forEach(targetingKey =>
+                        data.targeting[targetingKey] = bidObject[targetingKey]);
+      }
+
+      delete data['pubId'];
+      data.bidTimeout ? delete data.bidTimeout : null;
+      doubleclick(global, data);
```
Comment (lannka) [inline ads/aps.js]: @keithwrightbos @jasti as this is going to be deprecated Q1/2018, what should they do?

Comment (jasti) [reply ads/aps.js]: @AmazonPublisherServices for header bidding like support, please use RTC: https://github.com/ampproject/amphtml/blob/master/extensions/amp-a4a/rtc-documentation.md . CC @bradfrizzell 

The previous pattern to directly call DoubleClick is not supported anymore.

Comment (AmazonPublisherServices) [reply ads/aps.js]: Thank you @jasti and @lannka for reviewing this We can close this pull request as we are not going to be moving forward with our delayed fetch integration.

**Pull request closed without merging**